### PR TITLE
Stabilize `[const_]pointer_byte_offsets`

### LINKED ISSUE #96283: Tracking Issue for pointer_bytes_offsets
Feature gates: 

`#![feature(pointer_byte_offsets)]` 

`#![feature(const_pointer_byte_offsets)]`

This is a tracking issue for the pointer_byte_offsets raw pointer conveniences like `ptr.byte_add(offset)`

### Public API

```rust
impl<T: ?Sized> *const T {
    // feature gates `pointer_byte_offsets` and `const_pointer_byte_offsets`
    pub const unsafe fn byte_offset(self, count: isize) -> Self;
    pub const unsafe fn byte_add(self, count: usize) -> Self;
    pub const unsafe fn byte_sub(self, count: usize) -> Self;

    pub const fn wrapping_byte_offset(self, count: isize) -> Self;
    pub const fn wrapping_byte_add(self, count: usize) -> Self;
    pub const fn wrapping_byte_sub(self, count: usize) -> Self;

    pub const unsafe fn byte_offset_from<U: ?Sized>(self, origin: *const U) -> isize;
}

// ... and the same for` *mut T`
```

### Steps / History

- [x] Implementation: #95643
- [x] Making `byte_offset_from` more generic: #103489
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- Should these operations actually accomodate DSTs? This seems deeply semantically dubious/broken but idk, you can plausibly use them right if you are extremely careful and understand the implications of stacked borrows for slices/projections.

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Closes #96283
Awaiting FCP completion: https://github.com/rust-lang/rust/issues/96283#issuecomment-1735835331

r? libs-api

### CODE REVIEW

File: tests/ui/consts/miri_unleashed/ptr_arith.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // compile-flags: -Zunleash-the-miri-inside-of-you
-#![feature(core_intrinsics, pointer_byte_offsets)]
+#![feature(core_intrinsics)]
```
Comment (WaffleLapkin) [inline tests/ui/consts/miri_unleashed/ptr_arith.rs:2]: fyi: this test does not use either of those features :/

**Review by dtolnay — APPROVED** [review]

File: tests/ui/consts/miri_unleashed/ptr_arith.rs (hunk 1 above)
Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/ptr_arith.rs:2]: Ah, that's probably a silly copy-paste error then. Can you remove both feature flags while you are at it?

**Review by dtolnay — APPROVED** [review]
Thank you!

File: library/core/src/ptr/const_ptr.rs [hunk 2]
```diff
@@ -1045,8 +1049,9 @@ impl<T: ?Sized> *const T {
     /// leaving the metadata untouched.
     #[must_use]
     #[inline(always)]
-    #[unstable(feature = "pointer_byte_offsets", issue = "96283")]
-    #[rustc_const_unstable(feature = "const_pointer_byte_offsets", issue = "96283")]
+    #[stable(feature = "pointer_byte_offsets", since = "CURRENT_RUSTC_VERSION")]
+    #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "CURRENT_RUSTC_VERSION")]
+    #[rustc_allow_const_fn_unstable(set_ptr_value)]
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:1054]: Doesn't seem like @rust-lang/wg-const-eval got pinged for this? This exposes a const-intrinsic on stable, the WG should always be involved when that happens.

Comment (WaffleLapkin) [reply library/core/src/ptr/const_ptr.rs:1054]: I can't find any mentions of the WG either, so this looks like an overlook on my part 💀 

We should probably have automation for this kind of thing (i.e. ping the WG if the attributes are getting changed).

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:1054]: Yeah that would be nice, but seems hard -- I don't think we have any such automation for any case, currently.

Comment (WaffleLapkin) [reply library/core/src/ptr/const_ptr.rs:1054]: Yes, rustbot currently only supports pings based on file names, not diff contents. Although I feel like it shouldn't be too hard to add this functionality :thinking:

**Pull request merged**